crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    certified_update_producer::CertifiedUpdateProducer,
+    jwk_observer::JWKObserver,
+    network::IncomingRpcRequest,
+    types::{JWKConsensusMsg, ObservedUpdate, ObservedUpdateResponse},
+};
+use anyhow::{anyhow, bail, Result};
+use aptos_channels::{aptos_channel, message_queues::QueueStyle};
+use aptos_crypto::{bls12381::PrivateKey, SigningKey};
+use aptos_logger::{debug, error, info};
+use aptos_types::{
+    account_address::AccountAddress,
+    epoch_state::EpochState,
+    jwks::{
+        jwk::JWKMoveStruct, AllProvidersJWKs, Issuer, ObservedJWKs, ObservedJWKsUpdated,
+        ProviderJWKs, QuorumCertifiedUpdate, SupportedOIDCProviders,
+    },
+    validator_txn::{Topic, ValidatorTransaction},
+};
+use aptos_validator_transaction_pool::{TxnGuard, VTxnPoolState};
+use futures_channel::oneshot;
+use futures_util::{
+    future::{join_all, AbortHandle},
+    FutureExt, StreamExt,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Duration,
+};
+
+/// `JWKManager` executes per-issuer JWK consensus sessions
+/// and updates validator txn pool with quorum-certified JWK updates.
+pub struct JWKManager {
+    /// Some useful metadata.
+    my_addr: AccountAddress,
+    epoch_state: Arc<EpochState>,
+
+    /// Used to sign JWK observations before sharing them with peers.
+    consensus_key: Arc<PrivateKey>,
+
+    /// The sub-process that collects JWK updates from peers and aggregate them into a quorum-certified JWK update.
+    certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+
+    /// When a quorum-certified JWK update is available, use this to put it into the validator transaction pool.
+    vtxn_pool: VTxnPoolState,
+
+    /// The JWK consensus states of all the issuers.
+    states_by_issuer: HashMap<Issuer, PerProviderState>,
+
+    /// Whether a CLOSE command has been received.
+    stopped: bool,
+
+    qc_update_tx: Option<aptos_channel::Sender<(), QuorumCertifiedUpdate>>,
```

Comment:
Not sure if we should `aptos_channel` everywhere. Seems many places we only need `unbounded_channel` or `oneshot_channel`. What do you recommend @zekun000 ?

Reply:
isn't it in general safe to avoid *unbounded* channels?

Reply:
I saw it is used many places in consensus..

Reply:
i have updated to `qc_update_tx: Option<aptos_channel::Sender<Issuer, QuorumCertifiedUpdate>>,` and setup the queue to allow at most 1 msg per key. Which I believe makes a bit more sense: in case there are 2+ sessions finished for the same <issuer, version>, this setup helps us drop the result of the older one which is usually desired.

what's the concern of using aptos_channel in the first place?